### PR TITLE
feat: enhance naming for vpc resources and fix prefix checks

### PR DESCRIPTION
**Pull request by rajatagarwal-ibm**
### Description

Addresses:
* https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone-vpc/issues/678
* https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone-vpc/issues/677

### Release required?

- [ ] No release
- [ ] Patch release (`x.x.X`)
- [x] Minor release (`x.X.x`)
- [ ] Major release (`X.x.x`)

### Run the pipeline

If the CI pipeline doesn't run when you create the PR, the PR requires a user with GitHub collaborators access to run the pipeline.

Run the CI pipeline when the PR is ready for review and you expect tests to pass. Add a comment to the PR with the following text:

```
/run pipeline
```

### Checklist for reviewers

- [ ] If relevant, a test for the change is included or updated with this PR.
- [ ] If relevant, documentation for the change is included or updated with this PR.

### For mergers

- Use a conventional commit message to set the release level. Follow the [guidelines](https://terraform-ibm-modules.github.io/documentation/#/merging.md).
- Include information that users need to know about the PR in the commit message. The commit message becomes part of the GitHub release notes.
- Use the **Squash and merge** option.

### CODE REVIEW

**Comment by rajatagarwal-ibm** [comment]
Ran every test locally and also ran hub-spoke-manual-resolver apply (doesn't have test) - everything worked fine.

**Comment by rajatagarwal-ibm** [comment]
Need to skip the upgrade test as there is a change of the vpc name in the `main.tf`

**Comment by ocofaigh** [comment]
@rajatagarwal-ibm It seems you added several things since skipping the upgrade test so I'm not comfortable with it being skipped. Can you unskip it (use commit with `UNSKIP UPGRADE TEST`) and paste the results. We can then make a call on skipping it if no breaking changes are identified.

File: examples/existing_vpc/variables.tf [hunk 1]
```diff
@@ -9,10 +9,17 @@ variable "region" {
...
+
 variable "vpc_id" {
   description = "The ID of the VPC where the VSI will be created."
   type        = string
 }
+
+variable "public_gateway_name" {
+  description = "The name of the public gateway"
+  type        = string
+}
```
Comment (ocofaigh) [inline examples/existing_vpc/variables.tf:21]: Instead of adding a new variable, can you just hard code the variable values in the main.tf of the existing_vpc example main.tf ?

File: examples/existing_vpc/main.tf [hunk 2]
```diff
@@ -16,5 +16,6 @@ module "slz_vpc" {
   create_vpc          = false
   existing_vpc_id     = var.vpc_id
   create_subnets      = false
+  public_gateway_name = var.public_gateway_name
```
Comment (ocofaigh) [inline examples/existing_vpc/main.tf:19]: hard code a name here instead of using variable value (perhaps use the var.prefix in the name)

File: variables.tf [hunk 3]
```diff
@@ -3,11 +3,47 @@
 ##############################################################################
 
 variable "name" {
-  description = "The name to give the newly provisioned VPC. Only used if 'create_vpc' is true."
+  description = "The name to give the provisioned VPC. Only used if 'create_vpc' is true."
```
Comment (ocofaigh) [inline variables.tf:6]: hmm, name is also used for other resources, not just the VPC. I would update this to:

"The string to use for the naming of VPC resources."

I think you should also then add a new variable called `vpc_name` which can be used to explicitly name the VPC.

File: main.tf [hunk 4]
```diff
@@ -34,13 +34,27 @@ locals {
...
+}
+
 ##############################################################################
 # Create new VPC
 ##############################################################################
 
 resource "ibm_is_vpc" "vpc" {
   count                       = var.create_vpc == true ? 1 : 0
-  name                        = var.prefix != null ? "${var.prefix}-${var.name}-vpc" : "${var.name}-vpc"
+  name                        = var.prefix != null ? "${var.prefix}-${var.name}-vpc" : var.name
```
Comment (ocofaigh) [inline main.tf:57]: As per my other comment, var.name is used for other VPC resource naming. I would suggest you create a new variable called `var.vpc_name` (defaults to null) and regardless of the value of `var.prefix`, if the user passes a value for `var.vpc_name` the code should use that. Otherwise use "${var.prefix}-${var.name}-vpc" unless prefix is null, in which case use "${var.name}-vpc"

File: main.tf [hunk 5]
```diff
@@ -100,7 +114,7 @@ resource "ibm_is_vpc" "vpc" {
 resource "ibm_is_vpc_dns_resolution_binding" "vpc_dns_resolution_binding_id" {
   count = (var.enable_hub == false && var.enable_hub_vpc_id) ? 1 : 0
 
-  name   = "${var.prefix}-dns-binding"
+  name   = var.prefix != null ? "${var.prefix}-dns-binding" : var.dns_binding_name
```
Comment (ocofaigh) [inline main.tf:117]: Regardless of the value of `var.prefix`, if the user passes a value for `var.dns_binding_name` the code should use that. Otherwise use "${var.prefix}-${var.name}-dns-binding" unless prefix is null, in which case use "${var.name}-dns-binding"

File: main.tf [hunk 6]
```diff
@@ -109,26 +123,17 @@ resource "ibm_is_vpc_dns_resolution_binding" "vpc_dns_resolution_binding_id" {
 
 resource "ibm_is_vpc_dns_resolution_binding" "vpc_dns_resolution_binding_crn" {
   count  = (var.enable_hub == false && var.enable_hub_vpc_crn) ? 1 : 0
-  name   = "${var.prefix}-dns-binding"
+  name   = var.prefix != null ? "${var.prefix}-dns-binding" : var.dns_binding_name
```
Comment (ocofaigh) [inline main.tf:126]: Regardless of the value of `var.prefix`, if the user passes a value for `var.dns_binding_name` the code should use that. Otherwise use "${var.prefix}-${var.name}-dns-binding" unless prefix is null, in which case use "${var.name}-dns-binding"

File: main.tf [hunk 7]
```diff
@@ -109,26 +123,17 @@ resource "ibm_is_vpc_dns_resolution_binding" "vpc_dns_resolution_binding_id" {
...
-
-locals {
-  vpc_id = var.create_vpc ? ibm_is_vpc.vpc[0].id : var.existing_vpc_id
-}
-
 # Configure custom resolver on the hub vpc
 resource "ibm_resource_instance" "dns_instance_hub" {
   count             = var.enable_hub && !var.skip_custom_resolver_hub_creation && !var.use_existing_dns_instance ? 1 : 0
-  name              = "${var.prefix}-dns-instance"
+  name              = var.prefix != null ? "${var.prefix}-dns-instance" : var.dns_instance_name
```
Comment (ocofaigh) [inline main.tf:136]: Regardless of the value of `var.prefix`, if the user passes a value for `var.dns_instance_name` the code should use that. Otherwise use "${var.prefix}-${var.name}-dns-instance" unless prefix is null, in which case use "${var.name}-dns-instance"

File: main.tf [hunk 8]
```diff
@@ -137,7 +142,7 @@ resource "ibm_resource_instance" "dns_instance_hub" {
 
 resource "ibm_dns_custom_resolver" "custom_resolver_hub" {
   count             = var.enable_hub && !var.skip_custom_resolver_hub_creation ? 1 : 0
-  name              = "${var.prefix}-custom-resolver"
+  name              = var.prefix != null ? "${var.prefix}-custom-resolver" : var.dns_custom_resolver_name
```
Comment (ocofaigh) [inline main.tf:145]: Regardless of the value of `var.prefix`, if the user passes a value for `var.dns_custom_resolver_name` the code should use that. Otherwise use "${var.prefix}-${var.name}-custom-resolver" unless prefix is null, in which case use "${var.name}-custom-resolver"

File: main.tf [hunk 9]
```diff
@@ -193,7 +198,7 @@ resource "time_sleep" "wait_for_authorization_policy" {
 
 resource "ibm_is_vpc_routing_table" "route_table" {
   for_each                      = module.dynamic_values.routing_table_map
-  name                          = var.prefix != null ? "${var.prefix}-${var.name}-route-${each.value.name}" : "${var.name}-route-${each.value.name}"
+  name                          = var.prefix != null ? "${var.prefix}-${var.name}-route-${each.value.name}" : "${var.routing_table_name}-${each.value.name}"
```
Comment (ocofaigh) [inline main.tf:201]: I don't see any reason to add `var.routing_table_name` - you can achieve this naming with the current code using `var.name`. There can be multiple routing tables here, so we will always be appending `${each.value.name}` to the name here

File: main.tf [hunk 10]
```diff
@@ -228,7 +233,7 @@ locals {
 
 resource "ibm_is_public_gateway" "gateway" {
   for_each       = local.gateway_object
-  name           = var.prefix != null ? "${var.prefix}-${var.name}-public-gateway-${each.key}" : "${var.name}-public-gateway-${each.key}"
+  name           = var.prefix != null ? "${var.prefix}-${var.name}-public-gateway-${each.key}" : "${var.public_gateway_name}-${each.key}"
```
Comment (ocofaigh) [inline main.tf:236]: I don't see any reason to add `var.public_gateway_name` - you can achieve this naming with the current code using `var.name`. There can be multiple public gateways here, so we will always be appending `${each.value.name}` to the name here

File: variables.tf [hunk 11]
```diff
@@ -3,11 +3,47 @@
...
+  default     = null
+}
+
+variable "dns_custom_resolver_name" {
+  description = "The name to give the provisioned DNS custom resolver instance. Only used if 'enable_hub' is true and 'skip_custom_resolver_hub_creation'is false."
+  type        = string
+  default     = null
+}
+
+variable "routing_table_name" {
```
Comment (ocofaigh) [inline variables.tf:29]: as per above comment, I propose to delete this. There can be multiple routing tables so var.name (and var.prefix is not null) can be used for the naming

File: variables.tf [hunk 12]
```diff
@@ -3,11 +3,47 @@
...
+  description = "The name to give the provisioned Routing table."
+  type        = string
+  default     = null
+}
+
+variable "public_gateway_name" {
```
Comment (ocofaigh) [inline variables.tf:35]: as per above comment, I propose to delete this. There can be multiple public gateways so var.name (and var.prefix is not null) can be used for the naming

File: variables.tf [hunk 13]
```diff
@@ -3,11 +3,47 @@
...
+  description = "The name to give the provisioned VPC Public Gateway."
+  type        = string
+  default     = null
+}
+
+variable "vpc_flow_logs_name" {
+  description = "The name to give the provisioned VPC flow logs."
```
Comment (ocofaigh) [inline variables.tf:42]: After you update the logic in my other comment, The description here should be updated to: 

"The name to give the provisioned VPC flow logs. If no value is used, a name will be generated using the `name` and `prefix` (if using) input variables."

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@rajatagarwal-ibm See my comments please

File: tests/pr_test.go [hunk 14]
```diff
@@ -84,7 +84,7 @@ func TestRunNoPrefixExample(t *testing.T) {
 
 	// check if name is the same
 	outputs := terraform.OutputAll(options.Testing, options.TerraformOptions)
-	assert.Equal(t, testName+"-vpc", outputs["vpc_name"], "VPC name was altered, and should not have been.")
```
[the remark this answers is not in the document]
Comment (rajatagarwal-ibm) [reply tests/pr_test.go:87]: Yes, the assertion is working fine. I changed the assertion and it failed.

image

File: main.tf (hunk 5 above)
Comment (rajatagarwal-ibm) [reply main.tf:117]: Now changed it to the following. It will always check first if `dns_binding_name` is there or not, if no then check with prefix or name. 

```
name   = var.dns_binding_name != null ? var.dns_binding_name : (var.prefix != null ? "${var.prefix}-dns-binding" : "${var.name}-dns-binding")
```

File: main.tf (hunk 9 above)
Comment (rajatagarwal-ibm) [reply main.tf:201]: As the objective of the task is to give freedom to the user to name it anything they like, I have added these as variables. If they don't provide the value then it will pick up either prefix or name.

File: variables.tf (hunk 3 above)
Comment (rajatagarwal-ibm) [reply variables.tf:6]: This module is being consumed in the SLZ module. Any changes which we make here with `name` will impact that too. I don't think so that's needed. Hence my proposal is I can rename the description to the following:

```
The string to use for the naming of VPC and its resources.
```

File: main.tf (hunk 9 above)
Comment (ocofaigh) [reply main.tf]: But how are you going to handle the naming of multiple routing tables with 1 string variable. You dont want them all named the same, so hence you will need to use `${each.value.name}` in the naming

File: main.tf (hunk 10 above)
Comment (ocofaigh) [reply main.tf]: same as https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone-vpc/pull/685#discussion_r1426907203

File: main.tf [hunk 15]
```diff
@@ -34,13 +34,27 @@ locals {
...
+}
+
 ##############################################################################
 # Create new VPC
 ##############################################################################
 
 resource "ibm_is_vpc" "vpc" {
   count                       = var.create_vpc == true ? 1 : 0
-  name                        = var.prefix != null ? "${var.prefix}-${var.name}-vpc" : "${var.name}-vpc"
+  name                        = var.name != null ? var.name : "${var.prefix}-vpc"
```
Comment (vburckhardt) [inline main.tf:57]: This is going to break of name and prefix are both set to null. Perhaps add a check to ensure that one of them is set.

File: main.tf [hunk 16]
```diff
@@ -100,16 +114,17 @@ resource "ibm_is_vpc" "vpc" {
...
     id = var.hub_vpc_id # Target VPC ID
   }
 }
 
 resource "ibm_is_vpc_dns_resolution_binding" "vpc_dns_resolution_binding_crn" {
-  count  = (var.enable_hub == false && var.enable_hub_vpc_crn) ? 1 : 0
-  name   = "${var.prefix}-dns-binding"
+  count = (var.enable_hub == false && var.enable_hub_vpc_crn) ? 1 : 0
+
+  name   = var.dns_binding_name != null ? var.dns_binding_name : (var.prefix != null ? "${var.prefix}-dns-binding" : "${var.name}-dns-binding")
```
Comment (vburckhardt) [inline main.tf:127]: Using coalesce may make this easier to read. Also I'd suggest adding a comment above.

**Review by vburckhardt — COMMENTED** [review]
Hi Rajat, some comments

File: examples/existing_vpc/main.tf (hunk 2 above)
Comment (rajatagarwal-ibm) [reply examples/existing_vpc/main.tf:19]: There is no var.prefix in this example. In the tests, I am passing like `fmt.Sprintf("%s-public-gateway", prefix),`

File: main.tf [hunk 17]
```diff
@@ -194,8 +212,9 @@ resource "time_sleep" "wait_for_authorization_policy" {
 ##############################################################################
 
 resource "ibm_is_vpc_routing_table" "route_table" {
-  for_each                      = module.dynamic_values.routing_table_map
-  name                          = var.prefix != null ? "${var.prefix}-${var.name}-route-${each.value.name}" : "${var.name}-route-${each.value.name}"
+  for_each = module.dynamic_values.routing_table_map
+  # Use var.routing_table_name if not null, otherwise, use var.prefix or var.name.
+  name                          = var.routing_table_name != null ? "${var.routing_table_name}-${each.value.name}" : (var.prefix != null ? "${var.prefix}-route-${each.value.name}" : "${var.name}-route-${each.value.name}")
```
Comment (ocofaigh) [inline main.tf:217]: If the consumer passed a value for `prefix` and for `name` (and no value for `routing_table_name`), then I think we will want the routing table name name to be called `${var.prefix}-${var.name}-route-${each.value.name}`. Currently logic does not take that into consideration.

File: main.tf [hunk 18]
```diff
@@ -101,45 +122,42 @@ resource "ibm_is_vpc" "vpc" {
...
   plan              = var.dns_plan
 }
 
 resource "ibm_dns_custom_resolver" "custom_resolver_hub" {
-  count             = var.enable_hub && !var.skip_custom_resolver_hub_creation ? 1 : 0
-  name              = "${var.prefix}-custom-resolver"
+  count = var.enable_hub && !var.skip_custom_resolver_hub_creation ? 1 : 0
+
+  # Use var.dns_custom_resolver_name if not null, otherwise, use var.prefix or var.name.
+  name              = coalesce(var.dns_custom_resolver_name, "${var.prefix != null ? var.prefix : var.name}-custom-resolver")
```
Comment (ocofaigh) [inline main.tf:160]: If the consumer passed a value for `prefix` and for `name` (and no value for `dns_custom_resolver_name`), then I think we will want the dns resolver name name to be called `${var.prefix}-${var.name}-custom-resolver`. Currently logic does not take that into consideration.

File: main.tf [hunk 19]
```diff
@@ -101,45 +122,42 @@ resource "ibm_is_vpc" "vpc" {
...
-}
-
 # Configure custom resolver on the hub vpc
 resource "ibm_resource_instance" "dns_instance_hub" {
-  count             = var.enable_hub && !var.skip_custom_resolver_hub_creation && !var.use_existing_dns_instance ? 1 : 0
-  name              = "${var.prefix}-dns-instance"
+  count = var.enable_hub && !var.skip_custom_resolver_hub_creation && !var.use_existing_dns_instance ? 1 : 0
+
+  # Use var.dns_instance_name if not null, otherwise, use var.prefix or var.name.
+  name              = coalesce(var.dns_instance_name, "${var.prefix != null ? var.prefix : var.name}-dns-instance")
```
Comment (ocofaigh) [inline main.tf:149]: If the consumer passed a value for `prefix` and for `name` (and no value for `dns_instance_name `), then I think we will want the dns instance name name to be called `${var.prefix}-${var.name}-dns-instance`. Currently logic does not take that into consideration.

File: main.tf [hunk 20]
```diff
@@ -101,45 +122,42 @@ resource "ibm_is_vpc" "vpc" {
...
   }
 }
 
 resource "ibm_is_vpc_dns_resolution_binding" "vpc_dns_resolution_binding_crn" {
-  count  = (var.enable_hub == false && var.enable_hub_vpc_crn) ? 1 : 0
-  name   = "${var.prefix}-dns-binding"
+  count = (var.enable_hub == false && var.enable_hub_vpc_crn) ? 1 : 0
+
+  # Use var.dns_binding_name if not null, otherwise, use var.prefix or var.name.
+  name   = coalesce(var.dns_binding_name, "${var.prefix != null ? var.prefix : var.name}-dns-binding")
```
Comment (ocofaigh) [inline main.tf:137]: If the consumer passed a value for `prefix` and for `name` (and no value for `dns_binding_name `), then I think we will want the dns binding name name to be called `${var.prefix}-${var.name}-dns-binding`. Currently logic does not take that into consideration.

File: main.tf [hunk 21]
```diff
@@ -101,45 +122,42 @@ resource "ibm_is_vpc" "vpc" {
 resource "ibm_is_vpc_dns_resolution_binding" "vpc_dns_resolution_binding_id" {
   count = (var.enable_hub == false && var.enable_hub_vpc_id) ? 1 : 0
 
-  name   = "${var.prefix}-dns-binding"
+  # Use var.dns_binding_name if not null, otherwise, use var.prefix or var.name.
+  name   = coalesce(var.dns_binding_name, "${var.prefix != null ? var.prefix : var.name}-dns-binding")
```
Comment (ocofaigh) [inline main.tf:126]: If the consumer passed a value for `prefix` and for `name` (and no value for `dns_binding_name `), then I think we will want the dns binding name name to be called `${var.prefix}-${var.name}-dns-binding`. Currently logic does not take that into consideration.

**Review by ocofaigh — CHANGES_REQUESTED** [review]
@rajatagarwal-ibm I left some comments - maybe we need to jump on a call to discuss how we want this to look as its starting to get confusing now to support all the use cases

File: main.tf [hunk 22]
```diff
@@ -34,17 +40,31 @@ locals {
...
+}
+
 ##############################################################################
 # Create new VPC
 ##############################################################################
 
 resource "ibm_is_vpc" "vpc" {
   count          = var.create_vpc == true ? 1 : 0
-  name           = var.prefix != null ? "${var.prefix}-${var.name}-vpc" : "${var.name}-vpc"
+  name           = "${var.prefix != null && var.name != null ? "${var.prefix}-${var.name}" : var.prefix != null ? var.prefix : var.name}-vpc"
```
Comment (ocofaigh) [inline main.tf:63]: This logic still does not provide the ability to have full control over the VPC name - suggest to add optional `vpc_name` variable

Comment (ocofaigh) [inline main.tf:63]: `var.prefix != null && var.name != null ? "${var.prefix}-${var.name}-vpc" : var.prefix != null ? "${var.prefix}-vpc" : var.name`

File: variables.tf [hunk 23]
```diff
@@ -55,6 +43,58 @@ variable "access_tags" {
...
+##############################################################################
+
+variable "prefix" {
+  description = "The value that you would like to prefix to the name of the resources provisioned by this module. Explicitly set to null if you do not wish to use a prefix."
+  type        = string
+  default     = null
+}
+
+variable "name" {
+  description = "The string to use for the naming of VPC, when var.create_vpc is true. This string is also used as a prefix for the naming of VPC resources."
```
Comment (ocofaigh) [inline variables.tf:57]: If prefix value is not passed, then the value of this variable is exactly what the VPC will be called.

File: README.md [hunk 24]
```diff
@@ -58,6 +58,10 @@ module vpc {
 }
 ```
 
+### Resource naming
+
+The module automatically generates names for the all provisioned VPC resources using the `var.prefix` input variable. You can selectively override this behavior by giving explicit names through the following input variables: `dns_binding_name`, `dns_instance_name`, `dns_custom_resolver_name`, `routing_table_name`, `public_gateway_name`, and `vpc_flow_logs_name`.
```
Comment (ocofaigh) [inline README.md:63]: var.name (for VPC name)

File: dynamic_values.tf [hunk 25]
```diff
@@ -4,7 +4,7 @@
 
 module "dynamic_values" {
   source               = "./dynamic_values"
-  prefix               = var.prefix != null ? "${var.prefix}-${var.name}" : var.name
+  prefix               = var.prefix != null ? "${var.prefix}-vpc" : var.name
```
[the remark this answers is not in the document]
Comment (ocofaigh) [reply dynamic_values.tf:7]: `var.prefix != null && var.name != null ? "${var.prefix}-${var.name}" : var.prefix != null ? "${var.prefix}" : var.name`

[the remark this answers is not in the document]
Comment (ocofaigh) [reply dynamic_values.tf:7]: ^^ try this and paste the terraform plan that is generated

File: main.tf [hunk 26]
```diff
@@ -34,17 +40,31 @@ locals {
...
   resource_group = var.resource_group_id
   classic_access = var.classic_access
   # address prefix is set to auto only if no address prefixes NOR any subnet is passed as input
-  address_prefix_management   = (length([for prefix in values(coalesce(var.address_prefixes, {})) : prefix if prefix != null]) != 0) || (length([for subnet in values(coalesce(var.subnets, {})) : subnet if subnet != null]) != 0) ? "manual" : null
+  address_prefix_management   = length([for prefix in values(coalesce(var.address_prefixes, {})) : prefix if prefix != null]) != 0 ? "manual" : null
```
Comment (ocofaigh) [inline main.tf:67]: why is this change needed?

File: variables.tf [hunk 27]
```diff
@@ -55,6 +43,58 @@ variable "access_tags" {
...
+  type        = string
+  default     = null
+}
+
+variable "dns_binding_name" {
+  description = "The name to give the provisioned VPC DNS resolution binding. If not set, the module generates a name based on the var.prefix variable."
```
Comment (ocofaigh) [inline variables.tf:63]: var.prefix and var.name variables

Comment (ocofaigh) [reply variables.tf:63]: repeat for others

File: variables.tf [hunk 28]
```diff
@@ -55,6 +43,58 @@ variable "access_tags" {
...
+}
+
+variable "dns_custom_resolver_name" {
+  description = "The name to give the provisioned DNS custom resolver instance. If not set, the module generates a name based on the var.prefix variable."
+  type        = string
+  default     = null
+}
+
+variable "routing_table_name" {
+  description = "The name to give the provisioned routing table. If not set, the module generates a name based on the var.prefix variable."
```
Comment (ocofaigh) [inline variables.tf:81]: routing tables

File: variables.tf [hunk 29]
```diff
@@ -55,6 +43,58 @@ variable "access_tags" {
...
+  type        = string
+  default     = null
+}
+
+variable "public_gateway_name" {
+  description = "The name to give the provisioned VPC public gateway. If not set, the module generates a name based on the var.prefix variable."
```
Comment (ocofaigh) [inline variables.tf:87]: public gateways

**Comment by rajatagarwal-ibm** [comment]
Created https://github.com/terraform-ibm-modules/terraform-ibm-landing-zone-vpc/pull/697 since we can no longer bypass the upgrade test in that version.

This can be closed
